proposed/caller-unsafe.md

```diff
@@ -0,0 +1,60 @@
+
+# Annotating unsafe code with RequiresUnsafe
+
+## Background
+
+C# has had the `unsafe` feature since 1.0. There are two different syntaxes for the feature: a block syntax that can be used inside methods and a modifier that can appear on members and types. The original semantics only concern pointers. An error is produced if a variable of pointer type appears outside an unsafe context. For the block syntax, this is anywhere inside the block; for members this is inside the member; for types this is anywhere inside the type. Pointer operations are not fully validated by the type system, so this feature is useful at identifying areas of code needing more validation. Unsafe has subsequently been augmented to also turn off lifetime checking for ref-like variables, but the fundamental semantics are unchanged -- the `unsafe` context serves only to avoid an error that would otherwise occur.
+
+While existing `unsafe` is useful, it is limited by only applying to pointers and ref-like lifetimes. Many methods may be considered unsafe, but the unsafety may not be related to pointers or ref-like lifetimes. For example, almost all methods in the System.RuntimeServices.CompilerServices.Unsafe class has the same safety issues as pointers, but do not require an `unsafe` block. The same is true of the System.RuntimeServices.CompilerServices.Marshal class.
+
+## Goals
+
+The existing unsafe system does not clearly identify which methods need hand-verification to use safely, and it doesn't indicate which methods claim to provide that verification (and produce a safe API).
+
+We want to achieve all of the following goals:
+
+* Clearly annotate which methods require extra verification to use safely
+* Clearly identify in source code which methods are responsible for performing extra verification
+* Provide an easy way to identify in source code if a project doesn't use unsafe code
+
+## Proposal
+
+We need to be able to annotate code as unsafe, even if it doesn't use pointers.
```

Comment:
This is the critical part.

Reply:
Assuming the rules are going to be the same for everyone (BCL / SDKs / everybody else), wouldn't we need the opposite as well?

> We need to be able to annotate code as safe, _even if it uses pointers._

Otherwise it seems to me that this will become viral (similar to the trimming attributes), where a lot of APIs are considered unsafe, because they eventually do something unsafe in some code path.

Reply:
That’s what unsafe does today. It lets you call unsafe things, but has no requirements on the callers.

Reply:
I'm a bit lost here.
Does this mean that to produce "safe" packages I'll have to get rid of all the code the uses Unsafe.As, CollectionsMarshal and others?
How can I "guaranty" to others that my implementation IS safe (even if I use a `ImmutableCollectionsLarshal.AsImmutableArray` in a single place in my code base)?
Will all my packages be categorized as "unsafe" because in one low-level assembly I do use a `Unsafe.As`?
Also, I fully agree with @rolfbjarne: FieldOffset are unsafe... 
So the question is: what code base remains on the "safe" side?

Reply:
> That’s what unsafe does today. It lets you call unsafe things, but has no requirements on the callers.

Ah, I (think I) see, this new attribute would need to be added manually, it won't be added automatically whenever someone uses unsafe code.

Reply:
> I'm a bit lost here. Does this mean that to produce "safe" packages I'll have to get rid of all the code the uses Unsafe.As, CollectionsMarshal and others? How can I "guaranty" to others that my implementation IS safe (even if I use a `ImmutableCollectionsLarshal.AsImmutableArray` in a single place in my code base)? Will all my packages be categorized as "unsafe" because in one low-level assembly I do use a `Unsafe.As`? Also, I fully agree with @rolfbjarne: FieldOffset are unsafe... So the question is: what code base remains on the "safe" side?

Does this doc mention marking packages as unsafe? I don't see it

Reply:
@olivier-spinelli -- Fair question. As @EgorBo  says, that's not what is being communicated. In terms of helping people understand what we're thinking, this was probably a poor doc to go first. Expect some higher-level docs that better explain a plan.

Reply:
There is also a widely-seen formalism to disallow `unsafe` syntax totally, instead using the `Marshal` family which is inefficient and typically more unsafe. We should provide a path for moving people away from that.